Daily CRON job to upload a copy of our dataset to Vercel blob storage

ob6160 opened the pull request:
## What does this change?

- We'd like to provide a copy of our dataset to those who ask. This is the first piece of the puzzle, ensuring that we extract, format, and upload a copy of our dataset which we can then link users to.
- This implements a new API route, protected by a `CRON_SECRET` that is triggered once a day by a Vercel scheduled job configred in vercel.json.
- The script fetches data directly from the database, uploads it to Vercel blob storage, and then tidies up the storage - leaving just the latest copy of the dataset.

Following this change will be a piece of work to create a UI for users to download this information

ob6160 commented:
> **question:** are the `prisma`, `vercel` & `pnpm` updates strictly related to this change? Could they be extracted ou to their own PR?

Prisma update was needed to give me the exclude functionality. Vercel + pnpm were just convenient to do at the same time iirc.

ob6160 commented on src/pages/api/cron/daily_export.page.ts:
```diff
@@ -0,0 +1,155 @@
+    let fetchedCount = 0;
+    let hasNextBatch = true;
+
+    for (
+      let batchNumber = 0;
+      hasNextBatch && batchNumber < batchLimit;
+      batchNumber++
+    ) {
```
I’ll have a go..

ob6160 replied:
I've pushed a change that uses an async generator, what d'you think? :)

mxdvl replied:
Force-pushing makes it much harder to review the changes, but in any case the key benefit of async generators is that you can [simply use `Array.fromAsync` rather than imperative loops](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/fromAsync)!

ob6160 replied:
oh cool - will try

ob6160 replied:
latest push uses `Array.fromAsync`

mxdvl replied:
Nice, thanks for humouring, hope you like it. I think I'd go a step further and lean into the generator syntax:


```ts
/**
 * An async generator that yields toilet records.
 * This version uses a cursor-based approach to paginate through results.
 */
export async function* fetchAllToilets(
  options: {
    batchSize?: number;
    batchLimit?: number;
    filter?: { active?: boolean };
    orderBy?: { field: keyof ExportToilet; direction: 'asc' | 'desc' };
    onProgress?: (fetchedCount: number, batch: readonly ExportToilet[]) => void;
  } = {},
): AsyncGenerator<ExportToilet, void, unknown> {
  const {
    batchSize = 100,
    batchLimit = Infinity, // Fetch all batches by default
    filter = { active: true },
    orderBy = { field: 'created_at', direction: 'asc' },
    onProgress,
  } = options;

  let cursor: string | undefined;
  let fetchedCount = 0;

  while (true) {
    // Fetch the next batch of records using a cursor-based pagination
    const toiletsBatch = await prisma.toilets.findMany({
      where: filter,
      include: { areas: { select: { name: true, id: true } } },
      omit: { removal_reason: true, contributors: true },
      orderBy: { [orderBy.field]: orderBy.direction },
      take: batchSize,
      skip: cursor ? 1 : 0,
      ...(cursor && { cursor: { id: cursor } }),
    });

    fetchedCount += toiletsBatch.length;
    cursor = toiletsBatch.at(-1)?.id;

    // Report progress if a callback is provided
    onProgress?.(fetchedCount, toiletsBatch);

    // Yield the current items from the batch
    yield *toiletsBatch;

    // if no cursor was found we've reached the end
    if (!cursor) return;
  }
}
```

Hoolean replied:
Look out, with that last one the `batchLimit` condition has gone walkies! :rotating_light:

Hoolean replied:
(have just been lurking on the email thread - to join in the bikeshed fun though, maybe a do-while could kill the underlying infinite loop and check cursor and the batch limit at the same time :smile:)

mxdvl replied:
@Hoolean that's the beauty of generators: the absence of a cursor is sufficient to indicate that we've reached the end of the loos!

ob6160 replied:
thanks both! @mxdvl I've adopted your suggestion 💡 
[fdaa153](https://github.com/public-convenience-ltd/toiletmap/pull/1754/commits/fdaa1531a6362e49c0d262ba07acbf859b8c3962)

ob6160 commented on package.json:
```diff
@@ -204,6 +205,19 @@
       "axios@>=0.8.1 <0.28.0": ">=0.28.0"
-    }
+    },
+    "onlyBuiltDependencies": [
+      "@parcel/watcher",
+      "@prisma/client",
+      "@prisma/engines",
+      "@swc/core",
+      "core-js",
+      "core-js-pure",
+      "cypress",
+      "esbuild",
+      "prisma",
+      "sharp",
+      "supabase"
+    ]
```
new version of pnpm wants us to say if it's okay to run these during installation

mxdvl commented on src/pages/api/cron/daily_export.page.ts:
```diff
@@ -0,0 +1,133 @@
+      where: filter,
+      include: { areas: { select: { name: true, id: true } } },
+      omit: { removal_reason: true, contributors: true },
+      orderBy: { [orderBy.field]: orderBy.direction },
```
**issue:** [the docs suggest we need to order on the cursor](https://www.prisma.io/docs/orm/prisma-client/queries/pagination). Should we be simply hardcode it and not make this a parameter? It's relatively trivial to sort the array subsequently in JavaScript.

ob6160 replied:
good spot! I'll make this change

ob6160 replied:
Improved comments + fixed ordering added here: [fdaa153](https://github.com/public-convenience-ltd/toiletmap/pull/1754/commits/fdaa1531a6362e49c0d262ba07acbf859b8c3962)

mxdvl commented on src/pages/api/cron/daily_export.page.ts:
```diff
@@ -0,0 +1,133 @@
+  // Flatten the array of batches into a single array of toilets
+  const toilets = batches.flat();
+
+  const toiletsJson = JSON.stringify(toilets);
```
**question:** Would people not prefer a CSV?

Hoolean replied:
As a dev consumer I would probably prefer something typed and schema-able like JSON, but can imagine CSV being useful for people doing data analysis, in pandas or spreadsheets or otherwise; maybe we can spin out an issue to support both? :)

ob6160 replied:
I've added support for a CSV export here :)
[d11da7d](https://github.com/public-convenience-ltd/toiletmap/pull/1754/commits/d11da7d1a0659acd916cfbcee406ccaa02303773)

ob6160 commented:
@Hoolean @mxdvl as promised earlier here are examples of both exports + example output using the latest implementation:

```
Fetched 4000 toilets so far
Fetched 8000 toilets so far
Fetched 12000 toilets so far
Fetched 15386 toilets so far
Deleted old file: exports/toilets-1742150548070.json
Deleted old file: exports/toilets-1742150548070.csv
Uploading JSON export progress: 0%
Uploading CSV export progress: 0%
...
Uploading CSV export progress: 100%
Uploading JSON export progress: 100%
Successfully uploaded JSON export to blob at https://p02w6qqjlqmja4sk.public.blob.vercel-storage.com/exports/toilets-1742150867994-rQowge4aPpKkpbm1PauR10jrfnuvJK.json
Successfully uploaded CSV export to blob at https://p02w6qqjlqmja4sk.public.blob.vercel-storage.com/exports/toilets-1742150867994-GRVOBwb8I3HlH6CkMzHXCrLI7RbW8d.csv
```

mxdvl commented on src/pages/api/cron/daily_export.page.ts:
```diff
@@ -0,0 +1,188 @@
+      onProgress: (count) => console.log(`Fetched ${count} toilets so far`),
+    }),
+  );
+  const toilets = batches.flat();
```
**suggestion:** I think this is now unnecessary with the `yield*` syntax

mxdvl commented on src/pages/api/cron/daily_export.page.ts:
```diff
@@ -0,0 +1,188 @@
+    // Clean the export directory before uploading new files.
+    await cleanExportDirectory();
+
+    const timestamp = Date.now();
```
**suggestion:** use `new Date().toISOString().slice()` for YYYY-MM-DD readability.
